### ISSUE TITLE
Spack fails to generate correct modulefiles for module-specified externals

### ISSUE THREAD

**Issue by koysean** [LLNL/spack#35212]
### Steps to reproduce

Set up a Spack instance with a dependency such as `mpich` provided by external modules in `packages.yaml`. Install a package with that dependency and generate the modulefiles:
```console
$ spack install hypre^mpich
[...]
==> mpich@50.1 : has external module in [mpich/50.1, libfabric/1.2.3, ...]
[+] /path/to/mpich (external mpich-50.1-yuxbwgffkpgydwdllhfoyrrk35x52h36)
==> Installing hypre-2.26.0-zot76ybciz2jjqxn74qscbvl6z6h6abf
[...]
$ spack module tcl refresh
```

The generated `hypre` modulefile in `$spack/share/spack/modules` will have the following lines:
```
if { ![ is-loaded 'mpich/50.1-yuxbwgf' ] } {
    module load mpich/50.1-yuxbwgf
}
```
which tries to load the spack-generated mpich module. That module, in turn, only contains the environment modifications that are implied by the inferred prefix and spack's package-specific environment modifications (the same root cause as https://github.com/spack/spack/issues/35168).

In short, the generated modules for externals completely ignore the modules listed in `packages.yaml` which provide the external itself.

Expected behavior would be for the `hypre` module to either directly load the external `mpich` modules or for the spack-generated `mpich` module to directly load the external modules specified in `packages.yaml`.

### Information on your system

* **Spack:** 0.20.0.dev0 (75f1077b4beaffffaaac8b18ef98585f76a983d8)
* **Python:** 3.6.15
* **Platform:** linux-sles15-icelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by shahzebsiddiqui** [comment]
@koysean after several iterations of E4S deployment we just stop generating modulefiles via spack and just force users to adopt `spack load` due to the nature of how modulefiles can interact with external modulefiles, having to deal with module conflicts and custom suffix. 

Finally the explosion of modules caused lots of headache for user. Settting up the module configurations in spack can be tricky but it is doable.

**Comment by koysean** [comment]
Thanks for the insight. Unfortunately due to https://github.com/spack/spack/issues/35168 and some facility-specific reasons we need to stick with environment modules for our deployments. For the time being we can work around this issue but it will require manual intervention every time a new module is generated.

**Comment by alalazo** [comment]
@koysean What did you try to setup module configuration to blend well with your external module?

**Comment by koysean** [comment]
It's been a while since I had to deal with this since I've been using a fork off of https://github.com/spack/spack/pull/35596 for a while. I'll try my best to remember what I considered. To clarify, this is for a site installation of Spack.

One option was to forego loading dependent modules entirely (`autoload: none`). However, since this installation uses `runpath` to set library search paths, this was not an option as we would have no control over what dependent libraries would be loaded at runtime and put the onus on the user to set up the correct runtime environment.

Another option was to rely on the Spack-generated modulefiles entirely. Unfortunately, this sets up a confusing scenario for users where there are multiple conflicting modulefiles for the exact same thing (the original modulefile and the Spack-generated modulefile); it could also lead to false conflicts between modules that depend on different "versions" of what's actually the same library. Since the system software stack depends on vendor-provided modules, we can't commit to an entirely Spack-managed module stack.

There was also the option of maintaining a list of excluded modules for externals. This had two issues. First, the excluded modules would not be autoloaded, which means some modulefiles would load incomplete dependencies. Second, this requires manual maintenance of an excluded module list, which is particularly strenuous when working on test and development systems which may have frequent updates to the software stack. 

That's about what I can remember. Some other solutions involved putting together fragile scripts to emulate the above PR after-the-fact. Let me know if I can clarify any points or if there's something I hadn't considered.

**Comment by alalazo** [comment]
> It's been a while since I had to deal with this since I've been using a fork off of https://github.com/spack/spack/pull/35596 for a while.

You're right about the time it took, sorry. I'm trying to slowly, but steadily, reduce the amount of non-triaged bug reports.

**Comment by alalazo** [comment]
@koysean Did you try adding a new module template, like shown [here](https://spack-tutorial.readthedocs.io/en/latest/tutorial_modules.html#working-with-templates)? This might need some boilerplate, but lets you fine tune each module file. With the selection rules you can also apply the template to a set of modules.

**Comment by koysean** [comment]
@alalazo Thanks for the suggestion and I appreciate you working through these bug reports. I've finally gotten around to learning how module templates work for another use case.

I haven't tried applying module templates to solve this issue, but it seems non-trivial and I can't think of a complete solution which isn't API-breaking. Here are my thoughts on the different parts of the problem and where we can solve them:
- Spack generates modulefiles for externals: not solveable through templates afaik. This would require either a filter within the module classes (as I've implemented) or a new selection rule in `modules.yaml` which allows for selecting externals (which would be really useful in any case!).
- Spack's module writer doesn't know about external modules and won't autoload external modules: maybe solveable through templates, but requires a lot of logic which feels like it should be handled internally.
- Externals can have multiple modules: Partially solveable through templates?

I believe the first two points are solved through these changes, which aren't API breaking.
- https://github.com/spack/spack/pull/35596/files#diff-6d9ded14d84473e4e7159c8e5cdfb1829dae755124df2656ea52f93667027265R781-R789
- https://github.com/spack/spack/pull/35596/files#diff-2f260ea196ce41266c818be09d549f82b1bd9ce1013141e583edb96c0462b995R307-R311

Those changes also address part of the issue of having multiple modules for a single spec, since those changes should be sufficient for Spack to generate a correct `autoloads` list and only generate modules for Spack-managed specs.

The part that remains unaddressed is that `get_module()` is coded to return a single module per spec, which simply isn't valid once we're dealing with externals modules. This then breaks the `module loads` and `module find` commands since they both rely on `get_module()`. As far as I can tell, resolving that breaks the API. I guess we could handle external modules entirely separately from `get_module()`?

Let me know your thoughts and if I can do anything to help.
